### PR TITLE
Make Arkouda nilable-ready

### PR DESCRIPTION
**Pull request by bradcray**
This PR updates Arkouda to work in the nilable vs. non-nilable class world introduced by Chapel 1.20 and retires the need for the `--legacy-classes` flag (removing it from the Makefile).  It extends and replaces PR #152 which was an earlier approach for doing this.

To sanity check this, I ran `tests/check.py` against it, which passed.

The main effort here relates to the MultiTypeSymbolTable and its lookup() routine since lookup() used to return `nil` as a sentinel value when a lookup() failed, which would require propagating nilable classes throughout the code.  Instead of doing that, I converted lookup() to throw upon failure and changed routines that call lookup() to throw the error further upwards.  As a result, `arkouda_server.chpl` now does the main `select` statement in a `try` block so that these errors can be handled after the message has been processed.

One impact of this refactoring is that the procedure name which made the failing lookup() call is no longer captured, so does not get put into the error string.  That said, the result is much simpler in the code than a previous approach (PR #152) in which `try...catch` blocks were inserted around the lookup() calls to handle the errors closer to the failure.  That approach could, itself, be improved by support for deferred initialization as proposed in https://github.com/chapel-lang/chapel/issues/14271, so in the future, if we wanted to re-introduce the procedure names into the error message, this would be a way to do so a bit more cleanly, by catching the errors locally, or rethrowing them, incorporating the failing routine name.

Other than that change, the modification generally cleans up the code nicely, replacing a number of nil checks to guard against failed lookups into `throws` annotations and a `catch` block at the top level.  Moreover, some cases that were not guarded against by nil checks generated compiler errors with this new structure.  Changing them to `throw` results in safer code than what was on master before.

The other main changes here include:

* In arrayMsg(), I pushed most of the logic relating to reading an entry into a helper routine in order to be able to store it directly into `entry` as a non-nilable object.  This is a case that, in the future, could benefit from deferred initialization to keep the code closer to how it was originally written.

* Similarly, in readhdfMsg(), I pushed the logic related to computing an entry into a nested `computeEntry()` procedure in order to store it directly into `entry` as a non-nilable.

* Three new error classes are introduced:
  - `UndefinedSymbolError` which is thrown by lookup() on failure
  - `UnhandledDataTypeError` which is thrown by the helper in the first bullet above
  - `UnhandledHDFSDataTypeError` which is thrown by the helper in the second bullet above

* One new error string-generating routine is added: unknownError() which I use for the catch-all case in the outer try statement "just in case."  It's not particularly helpful, but better than halting...

In each of these cases, I'm not fully convinced that I've taken the best approach, though I think it's acceptable.  I'm definitely open to feedback.

I've generally tried not to change the indentation of the original code in order to keep the diff easy to read for the sake of review.  But once we have a PR that the team is relatively happy with, the code should be re-indented (either as part of this PR or a follow-on one).

### CODE REVIEW

**Comment by bradcray** [comment]
@mhmerrill: I'm stopping here for now.  This implements what we discussed on the phone today (and what @BryantLam and I were discussing on PR #152.

**Comment by mhmerrill** [comment]
@bradcray @BryantLam I want to understand why we have the symbol table declared as an array of a nil-able class.
Could we have dispensed with nil-able classes altogether here?
In the `deleteEntry` procedure we assign `nil` where we could have used potentially used `tab[name].clear()` to clear the shared reference in the symbol table.
Am I understanding things correctly here?
reference line: https://github.com/bradcray/arkouda/blob/98e78559dd5d145f33ef209fbf22a4c8ebaae6ff/src/MultiTypeSymbolTable.chpl#L120

**Comment by BryantLam** [comment]
> In the `deleteEntry` procedure we assign `nil` where we could have used potentially used `tab[name].clear()` to clear the shared reference in the symbol table.

I think this is an interesting case of what it means to be a nilable class type. Normally, I'd say that the correct type is a nilable class because it conveys usage intent to the developer, but that's not quite true here. The class is only nil for a little bit before you then remove the name from the domain.

In a way, the act of removing the name from the domain could be a signal to drop the entry / implicitly nil it for the user. But I'm not sure how to do the initial construction of the entry as a non-nil transaction if the type was a non-nilable class. Adding an entry into the domain would temporarily create a nil entry before the program fills it in.

Anyway, yeah you could just call `.clear()` and break the invariant of the type. While it is technically safe since the domain is immediately modified afterward, calling `.clear()` might not be future-proof.

**Comment by BryantLam** [comment]
@mhmerrill -- To clarify what I said, you're correct. `SymTab.tab` should be a container of non-nilable classes because the table will "always" hold non-nil entries. Only the `addEntry` and `deleteEntry` methods would ever observe the momentarily nil entry when the domain is modified.

**Comment by bradcray** [comment]
> I want to understand why we have the symbol table declared as an array of a nil-able class.

Tagging on to what Bryant said above, I believe that the deleteEntry() routine is not so much the problem as addEntry() is.  Specifically, deleteEntry() can be written to simply do `tD -= name;` and this should cause the corresponding table entry to be deleted and removed, so the user program could never see it in a nilable state (I didn't make this change simply because I was trying to minimize the diff with master, but making it works).

However, the addEntry() call is more problematic due to the decoupled nature of domains and arrays:  Since the current associative domain and array interface requires us to add the index to the domain in one step and assign the corresponding element of the array as a second distinct step, there is a period where the array element needs to hold its default value, yet non-nilable classes don't have one.

[side note: Doing this in two steps actually "works" today, but I believe this is a bug, not a feature, so would suggest we build Arkouda to rely on it.  I opened https://github.com/chapel-lang/chapel/issues/14367 against this behavior and proposing an extension to associative domains and arrays that would support it in a more principled way]

Now that there is a map type, another option would be to switch to using it, which is arguably simpler and all that arkouda wants/needs.  Since the map type takes a new key and value simultaneously, this does work today and ought to continue to work, though I'm noting that the implementation is working currently only because of the aforementioned https://github.com/chapel-lang/chapel/issues/14367.

So if it were of interest, I think I could change this PR (or file a follow-up one) to use a `map` type rather than the associative domain/array pair and have it map to non-nilable classes and that that should continue to work into the future.

File: src/GenSymIO.chpl [hunk 1]
```diff
@@ -226,27 +235,42 @@ module GenSymIO {
...
+    try {
+    var entry: shared GenSymEntry = computeEntry(dataclass);
+    var rname = st.nextName();
+    st.addEntry(rname, entry);
+    return try! "created " + st.attrib(rname);
+    } catch e: UnhandledHDFSDataTypeError {
+      return try! "Error: detected unhandled datatype code %i".format(dataclass);
+    } catch {
+      return "Unexpected error";
+    }
```
[the remark this answers is not in the document]
Comment (BryantLam) [reply src/GenSymIO.chpl]: Okay. I think these added lines were the ones removed from the block way below. The diff changes aren't so apparent that it was original code.

**Comment by bradcray** [comment]
I tried and failed to switch to a `map` type (arguably due to immaturity in the map type), so forked that off into its own issue for future consideration (#169).

So back to Mike's question, I think that yes, ultimately the MultiTypeSymboltable should store non-nilable objects, but my sense is that Chapel is not yet at the point where that's feasible, arguably due to nilability still being a fairly new feature.  I'd propose we go with something similar to the current PR for now and look into switching to non-nilability in the future (and would be happy to file another issue for that if others agree).

I view this and the switch to map as being good acid tests for these features.

**Pull request merged**